pkg/controller/discovery/auth/sar.go

```diff
@@ -0,0 +1,331 @@
+			}
+		}
+	}
+
+	return mAllowed, nil
+}
+
+//
+// Get whether authenticated
+func (r *RBAC) Authenticated() bool {
+	return r != nil && *r.authenticated
+}
+
+//
+// Get the user name for the token.
+func (r *RBAC) User() string {
+	if r.Authenticated() {
+		return r.tokenStatus.User.Username
+	}
+
+	return ""
+}
+
+//
+// Get the user ID for the token.
+func (r *RBAC) UID() string {
+	if r.Authenticated() {
+		return r.tokenStatus.User.UID
+	}
+
+	return ""
+}
+
+//
+// Get the user's groups.
+func (r *RBAC) Groups() []string {
+	if r.Authenticated() {
+		return r.tokenStatus.User.Groups
+	}
+
+	return []string{}
+}
+
+//
+// Get the extra fields for the token.
+func (r *RBAC) Extra() map[string]v1.ExtraValue {
+	extra := map[string]v1.ExtraValue{}
+	if r.Authenticated() {
+		for k, v := range r.tokenStatus.User.Extra {
+			extra[k] = append(
+				v1.ExtraValue{},
+				v...)
+		}
+	}
+
+	return extra
+}
+
+//
+// Do subject access review.
+func (r *RBAC) sar(request *Review) (bool, error) {
+	sar := v1.SubjectAccessReview{
+		Spec: v1.SubjectAccessReviewSpec{
+			ResourceAttributes: request,
+			Groups:             r.Groups(),
+			User:               r.User(),
+			UID:                r.UID(),
+			Extra:              r.Extra(),
+		},
+	}
+	err := r.Client.Create(context.TODO(), &sar)
+	if err != nil {
+		Log.Trace(err)
+		return false, err
+	}
+
+	return sar.Status.Allowed, nil
+}
+
+//
+// Authenticate the token.
+// Build the `subjectClient`.
+func (r *RBAC) authenticate() (bool, error) {
+	if r.authenticated != nil {
+		return *r.authenticated, nil
+	}
+	r.authenticated = pointer.BoolPtr(false)
+	tr := &auth.TokenReview{
+		Spec: auth.TokenReviewSpec{
+			Token: r.Token,
+		},
+	}
+	err := r.Client.Create(context.TODO(), tr)
+	if err != nil {
+		Log.Trace(err)
+		return false, err
+	}
+	if !tr.Status.Authenticated {
+		return false, err
+	}
+	r.tokenStatus = tr.Status
+	r.authenticated = pointer.BoolPtr(true)
+
+	return true, nil
+}
+
+//
+// Load project list (as needed).
+// When project API not supported, it will fall-back to
+// a SAR to determine if the subject has GET on all namespaces and
+// add the `ALL: true` entry.
+func (r *RBAC) loadProjects() {
+	if r.allowedProject != nil {
+		return
+	}
+	r.allowedProject = map[string]bool{}
+	subjectClient, err := r.subjectClient()
+	if err != nil {
+		return
+	}
+	list := &project.ProjectList{}
+	err = subjectClient.List(context.TODO(), nil, list)
+	if err != nil {
+		allowed, err := r.sar(&Review{
+			Resource: Namespace,
+			Verb:     GET,
+			Name:     ALL,
+		})
+		if err == nil && allowed {
+			r.allowedProject[ALL] = true
+		} else {
+			Log.Trace(err)
+		}
+		return
+	}
+	for _, p := range list.Items {
+		r.allowedProject[p.Name] = true
+	}
+}
+
+//
+// Build the subject client using the REST
+// configuration and token.
+func (r *RBAC) subjectClient() (client.Client, error) {
+	update := func(cfg *rest.Config) {
+		cfg.BearerToken = r.Token
+		cfg.Burst = 1000
```

Comment:
Agreed.

Reply:
hrm so AFAIK all the clients that use the config with these fields seem to use the same value, might be missing something. Thought it might make refactoring these constants later safer/easier

That said, if it is not convincing, not a merge blocker from my end!